runtime/compiler/optimizer/J9EstimateCodeSize.cpp

```diff
@@ -411,6 +413,54 @@ TR_J9EstimateCodeSize::adjustEstimateForMethodInvoke(TR_ResolvedMethod* method,
    return false;
    }
 
+bool
+TR_J9EstimateCodeSize::adjustEstimateForConstArgs(TR_CallTarget * target, int32_t& value, float factor)
+   {
+   static char *disableConstArgWeightReduction = feGetEnv("TR_disableConstArgWeightReduction");
+   if (disableConstArgWeightReduction || !target->_calleeSymbol) return false;
+
+   int32_t originalWeight = value;
+   TR_LinkHead<TR_ParameterMapping> argMap;
+   if (((TR_J9InlinerPolicy *)_inliner->getPolicy())->validateArguments(target,argMap))
+      {
+      for (TR_ParameterMapping* parm = argMap.getFirst(); parm; parm = parm->getNext())
+         {
+         int32_t interimWeight = value;
+         TR::Node * parmNode = parm->_parameterNode;
+         if (parmNode->getSymbolReference())
+            {
+            if (parmNode->getSymbolReference()->getSymbol()->isStatic()
+               && parmNode->getSymbolReference()->getSymbol()->castToStaticSymbol()->isConstString())
+               {
+               value *= factor;
+               heuristicTrace(tracer(),"Setting size from %d to %d because arg is constant string.", interimWeight, value);
+               }
+            else if (parmNode->getOpCodeValue() == TR::aload
+                  && parmNode->getSymbolReference()->getSymbol()->isConstObjectRef())
+               {
+               value *= factor;
+               heuristicTrace(tracer(),"Setting size from %d to %d because arg is const object ref.", interimWeight, value);
+               }
+            else if (parmNode->getOpCodeValue() == TR::aloadi
+                  && parmNode->getSymbolReference() == comp()->getSymRefTab()->findJavaLangClassFromClassSymbolRef())
```

Comment:
Is this automatically a sign of a constant class ? Shouldn't the child be a `loadaddr` static of a class in addition ?

Reply:
I guess there are two different kinds of benefits here : 
1) devirtualization of calls to `hashCode` and `equals` that are likely when going into a method such as `HashMap.get` 
and 
2) constant folding of call to `hashCode` and some simplification of other code related to equality such as length checks if the value is a constant. 

Obviously the latter is a bigger benefit, but the former is more likely to occur (i.e. there are probably more calls where we know that the argument is a j/l/Class than ones where we know the identity of the class too). This could argue for treating these two cases separately in terms of the scale factor applied (a more aggressive scaling when we know there are constants involved). Regardless of the scaling factor in use, if we believe that we need adjustments in this other place in the inliner code to make any such scheme actually kick in in practice, then we should make changes for the non-constant j/l/Class at this other place too.  That would be my suggestion then, use different scale factors for these cases and make changes in all the places required.

In theory, one could make a similar argument about distinguishing between knowing that we have a String and constant String value too. In actual fact, the boxed primitive classes ought to probably also be considered in a similar vein,  i.e. there is benefit in knowing that the object is an Integer for example and a greater benefit in knowing what the constant value of the Integer is. Maybe these cases ought to be handled too, in the spirit of completeness (if you prefer, via a different PR).

Reply:
Thank you for the detailed response! I updated the branch with the suggested changes in ECS, where we have a regular adjustment factor and less aggressive adjustment factor in place. The regular adjustment factor is used when we have constant strings, constant classes and integer loads, and the less aggressive adjustment factor is used if we can not confirm that they are constant. I would appreciate your feedback on my approach in determining when to apply which factor, and whether the less aggressive adjustment factor needs to be altered. In the meantime, I will be testing the changes.

Right now, I do not think we would benefit from applying the same changes to `applyArgumentHeuristics`.

Reply:
Thanks. Can you please clarify why we don't believe that we would benefit from applying the same changes to `applyArgumentHeuristics` ? I would have thought that was a way of ensuring that any adjustment factor (less or more aggressive) has the desired effect in the end. But maybe I'm missing some other key point.

Reply:
The weight adjustment being applied in `adjustEstimateForConstArgs` proposed in this PR encourages inlining callees we are interested in mainly through reducing the possibility for individual callees from being discarded because they previously went over the max size threshold by a small enough margin. The margin would have to be small enough such that the adjustment factor being applied lowers the weight just enough to go through the size checks. Those callees previously exceeded size thresholds either because of their actual size estimate obtained from ECS or through a blown up size estimate in cases such as when the callee method is already compiled at a higher opt level than the root method.

`applyArgumentHeuristics` on the other hand has less of an impact in determining whether individual callees get removed from inlining, and has more of an impact on how much budget an inlining candidate is consuming out of the total inlining budget. `applyArgumentHeuristics` is used very late in the call weighing phase, and the size that this function works with now as a result of this changeset would already have the weight reduction factor applied, and therefore already resulting in adjusted weights having a lower inlining budget consumption.

Therefore, I do not expect much benefit in terms of encouraging inlining of const/non-const strings, etc. from making the changes to `applyArgumentHeuristics`.

Do you still think it is worth going ahead with making the changes to `applyArgumentHeuristics` so that we have consistency in the size/weight adjustments made for the different kinds of args in the two different functions?

Reply:
I feel consistency is valuable, to avoid developers in the future having to 1) debug a problem that has been found previously with respect to different places in the inliner that must be scaled down and 2) understand that whole process enough to fix the right places to sort out the case (const or non const) that they are interested in to be more like the other case. I would prefer the right/all places are addressed consistently and the only difference being in the scale factor values applied.

Reply:
There are a few issues with trying to achieve consistency that I came across while working on this:

1. Weight adjustments are [only apply for opt level > warm](https://github.com/eclipse/omr/blob/e09e7b5f19fdc34bace6f1b938f302b73471a23c/compiler/control/OMROptions.cpp#L2659-L2660), using the following formula in most cases:

https://github.com/eclipse-openj9/openj9/blob/a579974d288caad0e5b994b608f5030249dc061d/runtime/compiler/optimizer/InlinerTempForJ9.cpp#L2632

2. When dealing with "const classes", the weight gets set to 10, rather than applying an adjustment factor. If the Class arg appears at the beginning, then the weight gets set to that and the adjustments are done on that value as we iterate through the other args. If the class arg appears at the end, then all the adjustment factor applications are discarded and simply set to 10. So we would get different results for the same method by just modifying the order of the args.

https://github.com/eclipse-openj9/openj9/blob/a579974d288caad0e5b994b608f5030249dc061d/runtime/compiler/optimizer/InlinerTempForJ9.cpp#L2640-L2645

3. Similarly for `TR::Address` nodes, the weight gets reset to a specific value.

https://github.com/eclipse-openj9/openj9/blob/a579974d288caad0e5b994b608f5030249dc061d/runtime/compiler/optimizer/InlinerTempForJ9.cpp#L2647-L2650

4. For each arg, we also check if the arg has known object info (it's an `if` instead of `else if`), and then set the weight to 10 and just exit the for loop. Not sure why this is done, as this just resets any weight adjustment that have been done throughout the call weighing phase, and set to a value that can possibly be very far from what the actual weight was. Should this also be done for `adjustEstimateForConstArgs`?

https://github.com/eclipse-openj9/openj9/blob/a579974d288caad0e5b994b608f5030249dc061d/runtime/compiler/optimizer/InlinerTempForJ9.cpp#L2653-L2660

5. There is a final adjustment done at the end, which subtracts numArgs * 4, resulting in negative weights pretty often. Should this also be done in `adjustEstimateForConstArgs`?

https://github.com/eclipse-openj9/openj9/blob/a579974d288caad0e5b994b608f5030249dc061d/runtime/compiler/optimizer/InlinerTempForJ9.cpp#L2665

Reply:
I feel that applying heuristics consistently at this spot and at `adjustEstimateForConstArgs` using the logic at this spot  could be as far as we go in this PR (i.e. I feel 4 and 5 are worth doing). I feel the issues that you pointed out, e.g. 2 and 3 seem legitimate to me and perhaps ought to be addressed separately. For 1, we could change the code to apply the heuristics at warm too but again, this expands the scope of this PR in a way that may not be needed right away. I feel we would need to do more extensive compile time study to do that sort of heuristic change since warm compiles are more numerous and also where a bulk of compile time goes.

Reply:
@vijaysun-omr addressed the last comment with my latest update